apis/v1alpha1/httproute_types.go

```diff
@@ -532,6 +546,45 @@ type HTTPRequestMirrorFilter struct {
 	Port PortNumber `json:"port"`
 }
 
+// HTTPRequestAuthorization defines configuration for the RequestAuthorization filter.
+type HTTPRequestAuthorization struct {
+	// URL indicates the address of the service used to perform authorization checks.
+	//
+	// There are two protocols supported:
+	// - grpc: In this mode, the authorization server should receive a grpc request of type <service.auth.v3.CheckRequest>.
+	//   It is expected for the authorization server to implement the Envoy ExtAuthz GRPC protocol:
+	//   https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/external_auth.proto
+	//
+	// - https: An HTTP request that will include all the Headers and Request Path, but no body content.
+	//   It is expected for the authorization server to implement the Envoy ExtAuthz HTTP protocol:
+	//   https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/external_auth.proto
+	//
+	URL string `json:"url"`
```

Comment:
How does configuration of how to reach this URl work? For example, if we need authentication, TLS, etc.

if it were a k8s service it would be "easy" since existing APIs defined here like BackendPolicy would apply. For a generic URL, I am not sure its easy.

Reply:
Yes, that's a good point. Thinking out loud, perhaps splitting the URL into a `service reference` and a `type` or `protocol` field to define the type of authorization (grpc/HTTP) could work?

Reply:
That could potentially work, I think some sort of reference is probably more consistent with the rest of the API as well. FWIW we recently went through lots of debate with https://docs.google.com/document/d/1V4mCQCw7mlGp0zSQQXYoBdbKMDnkPOjeyUb85U07iSI/ about this exact topic in Istio.

Reply:
Thanks, @howardjohn! I will review that document and try to find a better API design.

Reply:
Just getting to review this one now, Contour added an ExtensionService to cover the "service reference" part of this (see https://projectcontour.io/docs/v1.12.0/config/api/#projectcontour.io/v1alpha1.ExtensionService), and then an AuthorizationServer struct that holds the ExtensionService reference. (https://projectcontour.io/docs/v1.12.0/config/api/#projectcontour.io/v1.AuthorizationServer). Both the ExtensionService and AuthorizationServer have places to put config.

I'm not suggesting doing the same thing, but I think the idea of a service reference and a type or protocol field would be great.

Reply:
@youngnick It already has a protocol field, but for the service reference, we can follow the example of the other filters and replace URL with:

```go
ServiceName *string `json:"serviceName,omitempty"`
BackendRef *LocalObjectReference `json:"backendRef,omitempty"`
Port *PortNumber `json:"port,omitempty"`
```

Or add a new structure that allows us to cross the namespace boundary, with additional information on how to connect to the external service, like for example a TLS client certificate to use.